utils/mgmt_system/openstack.py

```diff
@@ -435,8 +444,22 @@ def deploy_template(self, template, *args, **kwargs):
                                            *args, **kwargs)
         self.wait_vm_running(kwargs['vm_name'], num_sec=timeout)
         if kwargs.get('floating_ip_pool', None):
-            ip = self.api.floating_ips.create(kwargs['floating_ip_pool'])
+            pool = kwargs['floating_ip_pool']
+            free_ips = sorted(
+                self.api.floating_ips.findall(fixed_ip=None, pool=pool), key=lambda ip: ip.ip)
+            # We maintain 1 floating IP as a protection against race condition
+            # I know it is bad practice, but I did not figure out how to prevent the race
+            # condition by openstack saying "Hey, this IP is already assigned somewhere"
+            if len(free_ips) > 1:
+                # There are 2 and more ips, so we will take the first one (eldest)
+                ip = free_ips[0]
+                logger.info("Reusing {} from pool {}".format(ip.ip, pool))
+            else:
+                # There is one or none, so create one.
+                ip = self.api.floating_ips.create(pool)
```

Comment:
Does this raise an exception if there is not enough room in the pool? If so, should we catch and rethrow?